sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala

```diff
@@ -88,11 +88,29 @@ private case class MsSqlServerDialect() extends JdbcDialect with NoLegacyJDBCErr
         case e: Predicate => e.name() match {
           case "=" | "<>" | "<=>" | "<" | "<=" | ">" | ">=" =>
             val Array(l, r) = e.children().map {
-              case p: Predicate => s"CASE WHEN ${inputToSQL(p)} THEN 1 ELSE 0 END"
+              case p: Predicate => inputToCaseWhenSQL(p)
               case o => inputToSQL(o)
             }
             visitBinaryComparison(e.name(), l, r)
-          case "CASE_WHEN" => visitCaseWhen(expressionsToStringArray(e.children())) + " = 1"
+          case "CASE_WHEN" =>
+            // Since MsSqlServer cannot handle boolean expressions inside
+            // a CASE WHEN, it is necessary to convert those to an IIF
+            // expression that will return 1 or 0 depending on the result.
+            // Example:
+            // In:  ... CASE WHEN a = b THEN c = d ...
+            // Out: ... CASE WHEN a = b THEN IIF(c = d, 1, 0) ...
+
+            // grouped turns Array[Expression] to Array[Array[Expression]]
+            // with a len of max 2 (final one will have only one)
+            val stringArray = e.children().grouped(2).flatMap { arr =>
+              arr.dropRight(1).map(inputToSQL) :+
+                (arr.last match {
+                  case p: Predicate if p.name() != "ALWAYS_TRUE" && p.name() != "ALWAYS_FALSE" =>
```

Comment:
can we say more about what's special for ALWAYS_TRUE/ALWAYS_FALSE?

Reply:
ALWAYS_TRUE gets translated to 1, and that is an int. If wrapped, it will fail with the same issue we are trying to fix.

Reply:
Should I make a comment?

Reply:
> ALWAYS_TRUE gets translated to 1

Where does it happen?

Reply:
```
  override def compileValue(value: Any): Any = value match {
    case booleanValue: Boolean => if (booleanValue) 1 else 0
    case other => super.compileValue(other)
  }
```

In MsSqlServer

Reply:
hmm, is it correct? What if the boolean is in a place that do need a boolean type, like a function that takes a boolean? This is unrelated to this PR though.

Reply:
problem is that MsSql cannot ask for a boolean as it cannot handle booleans. Any other DB will know how to translate it so we are ok. Only way this poses a problem is if we say CASE WHEN TRUE THEN which makes no sense so we are good